### DISCUSSION TITLE
[sitecore-jss-proxy] Configurable Loglevel

### DISCUSSION THREAD

**Discussion by monkey-ldb** [Sitecore/jss#1401] [Ideas]
## Is your feature request related to a problem? Please describe.
I only want to log error messages, no info/warn, but I can only configure debug=true/false.
There seems no other way to pass this parameter to the http proxy framework, which is used.
https://github.com/Sitecore/jss/blob/af8d48cf53ab1547d8e0566a78ab7fbfad7c54f0/packages/sitecore-jss-proxy/src/index.ts#L556

## Describe the solution you'd like
config parameter to pass the expected log level

## Notes
Would be good if this feature will be added also to older versions, like 18.0

**Comment by ambrauer** [comment]
Hi @monkey-ldb. This is a great suggestion. I've added this to our backlog to prioritize.

We also gladly accept PRs if you're willing. Check the [contributing guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md) as it has had recent updates and we'll be happy to merge it in.

**Comment by mjarosch** [comment]
Hi @ambrauer, I stumbled across this and was wondering if it is still on your backlog? Also, you mentioned accepting PRs, what is the rule about external libraries? Like using [winston](https://github.com/winstonjs/winston) to re-implement and provide more advanced logging features?

**Reply by ambrauer** [reply]
Hi @mjarosch! Still in our backlog, but this particular log level is applicable only to the [http-proxy-middleware](https://github.com/chimurai/http-proxy-middleware/tree/v0.21.0?tab=readme-ov-file#options) used for `@sitecore-jss/sitecore-jss-proxy` package (I'll update the discussion title to make that clear). In all other SDK packages, JSS has standardized on use of the [debug](https://www.npmjs.com/package/debug) module for logging (see docs [here](https://doc.sitecore.com/xmc/en/developers/jss/216/jss-xmc/debug-logging-in-jss-apps.html) for more details).

**Reply by mjarosch** [reply]
Hi @ambrauer, I am aware of the debug functionality. It has nice name-spacing, but is lacking the logging level and other advanced features.

**Reply by ambrauer** [reply]
@mjarosch We are pretty committed to debug at this point. The big draws here were the flexibility of name-spacing, the environment variable-based API, server/browser support, but also the general simplicity. I think log level could be implemented with debug (e.g. using a `DEBUG_LEVEL` variable) but so far have avoided introducing that level of complexity. Worth considering though. Beside log level, what other features would you be looking for?

**Reply by mjarosch** [reply]
@ambrauer The big missing feature is transports. We centralize our logs in an APM like AppInsights or Dynatrace. There isn't a great way to accomplish that with debug.

I do like the environment variable configuration, but Vercel requires a redeploy to pickup changes. Not a show stopper, but a big speed bump when troubleshooting. Although, being able to point your local at production experience edge data has been a game changer for debugging.

**Reply by ambrauer** [reply]
@mjarosch I should have mentioned one other advantage with debug is that because it does use the Web Standard [console](https://developer.mozilla.org/en-US/docs/Web/API/Console) API, this automatically works on edge runtime. This is a requirement for Next.js but also a general expectation as the industry moves to serverless.

My take on transports is that this should be at a lower level (working directly with stderr or stdout) with a log drain approach. Most deployment platforms provide this feature (e.g. [Vercel ](https://vercel.com/docs/observability/log-drains)and [Netlify](https://docs.netlify.com/monitor-sites/log-drains/?monitoring-providers=datadog)). I can see how this would be more of a challenge for self-hosting though.
